### PR TITLE
BP-1260 - Add a new panic handler

### PR DESCRIPTION
**Pull request by sp-marcel-hernandez**

### CODE REVIEW

File: grpcx/server.go [hunk 1]
```diff
@@ -119,3 +123,37 @@ func WithDiscardedCodes(codes ...codes.Code) func(*errorLogsOptions) {
...
+
+	return func() {
+		if r := recover(); r != nil {
+			l.Info(fmt.Sprintf("%v", r), logx.F("stack_trace", strings.Split(string(debug.Stack()), "\n")))
+			ro.exitFunc()
+		}
+	}
+}
+
+func SetExitFunction(exitFunc func()) func(*recoveryOptions) {
```
Comment (sp-jordi-forns) [inline grpcx/server.go:148]: s/SetExitFunction/withExitFunction

File: grpcx/server.go [hunk 2]
```diff
@@ -119,3 +123,37 @@ func WithDiscardedCodes(codes ...codes.Code) func(*errorLogsOptions) {
 		logsConfig.discardedCodes = append(logsConfig.discardedCodes, codes...)
 	}
 }
+
+type recoveryOptions struct {
+	exitFunc func()
+}
+
+func NewRecoveryHandler(l logx.Logger, options ...func(*recoveryOptions)) func() {
```
Comment (sp-jordi-forns) [inline grpcx/server.go:131]: s/NewRecoveryHandler/RecoveryHandler

File: grpcx/server.go [hunk 3]
```diff
@@ -119,3 +123,37 @@ func WithDiscardedCodes(codes ...codes.Code) func(*errorLogsOptions) {
...
+			ro.exitFunc()
+		}
+	}
+}
+
+func SetExitFunction(exitFunc func()) func(*recoveryOptions) {
+	return func(ro *recoveryOptions) {
+		ro.exitFunc = exitFunc
+	}
+}
```
Comment (sp-jordi-forns) [inline grpcx/server.go:152]: the functions above are not bounded to grpcx, so they should be in another package.

File: grpcx/server.go [hunk 4]
```diff
@@ -119,3 +123,37 @@ func WithDiscardedCodes(codes ...codes.Code) func(*errorLogsOptions) {
...
+
+func WithStructuredPanicLogs(recoveryHandler func()) grpc.UnaryServerInterceptor {
```
Comment (sp-jordi-forns) [inline grpcx/server.go:154]: you don't need the recoveryHandler arg, just call RecoveryHandler

File: grpcx/server_test.go [hunk 5]
```diff
@@ -104,6 +104,29 @@ func TestWithRequestResponseLogs(t *testing.T) {
...
+		panickingHandler := grpc.UnaryHandler(func(ctx context.Context, req interface{}) (interface{}, error) {
+			panic("Here's Johnny!")
+		})
+
+		_, err := interceptor(ctx, exampleRequest, info, panickingHandler)
+
+		a.NoError(err)
+		a.Contains(w.String(), "Here's Johnny!")
+	})
+}
```
Comment (sp-jordi-forns) [inline grpcx/server_test.go:128]: i'd focus the test in RecoveryHandler, WithStructuredPanicLogs does barely nothing

**Review by sp-jordi-forns — CHANGES_REQUESTED** [review]

File: grpcx/server.go (hunk 4 above)
Comment (sp-marcel-hernandez) [reply grpcx/server.go:154]: It's easier to test this way, though.

File: recovery/recovery_test.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package recovery
+
+import (
+	"bytes"
+	"github.com/socialpoint-labs/bsk/logx"
+	"github.com/stretchr/testify/assert"
```
Comment (sp-jordi-forns) [inline recovery/recovery_test.go:6]: also these 2 separate

File: grpcx/server.go [hunk 7]
```diff
@@ -119,3 +120,10 @@ func WithDiscardedCodes(codes ...codes.Code) func(*errorLogsOptions) {
 		logsConfig.discardedCodes = append(logsConfig.discardedCodes, codes...)
 	}
 }
+
+func WithStructuredPanicLogs(l logx.Logger, options ...recovery.Options) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		defer recovery.Handler(l, options...)()
```
Comment (sp-jordi-forns) [inline grpcx/server.go:126]: keep the recovery handler in a var, for more clarity

File: recovery/recovery_test.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+func TestWithRequestResponseLogs(t *testing.T) {
+	t.Parallel()
+	a := assert.New(t)
+
+	t.Run("panic handler serializes panic", func(t *testing.T) {
+		w := bytes.NewBufferString("")
+		logger := logx.New(logx.WriterOpt(w))
+		blankExitFunc := func() {
+			a.Contains(w.String(), "test panicking")
```
Comment (sp-jordi-forns) [inline recovery/recovery_test.go:18]: you could also raise a flag here and make sure this code was actually run by checking it in the end, same as you did in grpcx.

**Review by sp-jordi-forns — CHANGES_REQUESTED** [review]

**Review by sp-jordi-forns — APPROVED** [review]

**Pull request merged**